# 021-P01. Sensor Log File

*Construct one file path, write and append records, read the file, and validate the saved values.*

Read the problem statement and the following theory sections before running the notebook: Paths and Filesystem State, Text and Structured File Round Trips.

## Case Data

The problem supplies three ordered temperature readings and separates the initial write from the later arrival. The only intended artifact is `output/sensor_log.csv`.

## Method Application · Initial File Output

The first operation constructs one relative output path and replaces that exact file with a header and two rows.

In [1]:
from pathlib import Path

output_dir = Path("output")
output_dir.mkdir(exist_ok=True)
log_path = output_dir / "sensor_log.csv"

initial_text = (
    "timestamp,temperature_c\n"
    "2026-03-02T09:00,21.4\n"
    "2026-03-02T09:05,21.8\n"
)
written_characters = log_path.write_text(initial_text, encoding="utf-8")

print("Saved:", log_path)
print("Characters written:", written_characters)

Saved: output/sensor_log.csv
Characters written: 68


## Method Application · Focused Append Operation

The append block adds the third row and immediately displays the complete saved text.

In [2]:
with log_path.open(mode="a", encoding="utf-8") as log_file:
    log_file.write("2026-03-02T09:10,22.1\n")

print(log_path.read_text(encoding="utf-8"))

timestamp,temperature_c
2026-03-02T09:00,21.4
2026-03-02T09:05,21.8
2026-03-02T09:10,22.1



## Method Application · Restored Temperature Values

The parser reads the artifact, skips its header, and converts the saved temperature fields to numbers.

In [3]:
lines = log_path.read_text(encoding="utf-8").strip().splitlines()
temperatures_c = []

for line in lines[1:]:
    fields = line.split(",")
    temperatures_c.append(float(fields[1]))

print("Recovered temperatures (°C):", temperatures_c)

Recovered temperatures (°C): [21.4, 21.8, 22.1]


## Results · File Summary Values

The final cell reports count, range, and mean from the restored file values.

In [4]:
mean_temperature_c = sum(temperatures_c) / len(temperatures_c)

print("Readings:", len(temperatures_c))
print("Minimum temperature (°C):", min(temperatures_c))
print("Maximum temperature (°C):", max(temperatures_c))
print("Mean temperature (°C):", round(mean_temperature_c, 2))

Readings: 3
Minimum temperature (°C): 21.4
Maximum temperature (°C): 22.1
Mean temperature (°C): 21.77


## Result Analysis

### Applied Theory

Interpret the case with Paths and Filesystem State and Text and Structured File Round Trips.

### Computed Results

The initial write stores 68 characters in `output/sensor_log.csv`. The append produces a file with one header and three ordered rows. Reading the file back recovers temperatures `[21.4, 21.8, 22.1]`, with a count of 3, a minimum of 21.4 °C, a maximum of 22.1 °C, and a mean of 21.77 °C.

### Interpretation

The reloaded text verifies that append mode retained the initial rows and added the later reading at the end. The parsed summary uses the saved artifact rather than separate in-memory values, so it tests the complete round trip.

### Assumptions and Limits

Each rerun intentionally replaces the same exact file before appending, and the parser assumes one header followed by comma-separated timestamp and temperature fields. The three readings do not describe long-term sensor behavior.

### Next Topic

The next session begins with Structure Selection and Growth, which chooses storage from the operation a workflow requires.
